DietPi-Software | Add Config Creation function

nonnorm opened the pull request:
<!--
Before submitting a pull request:
- Please ensure the target branch is "dev" (active development): https://github.com/MichaIng/DietPi/tree/dev
- Please ensure changes have been tested and verified functional.
-->
**Status**: Ready

- [x] Test on VM

**Reference**: https://github.com/MichaIng/DietPi/issues/3850
(Closes #3850)

**Commit list/description**:
<!--
- DietPi-Config | Add "Fan control" option to "Performance Options"
![Screenshot](https://xxx.github.com/images/xxx.png)
- DietPi-Config | Add "Fan control" support for Odroid C2
- DietPi-Config | Syntax fix
-->
- DietPi-Globals | Add Config Creation function
- DietPi-Globals | CodeFactor Fixes
- DietPi-Globals | More CodeFactor Fixes
- DietPi-Globals | Even more CodeFactor fixes
- DietPi-Globals | Fix G_DIETPI_VERSION?
- DietPi-Software | Move to DietPi-Software
- DietPi-Software | Use CONFIG_CREATE in Bazarr, Sonarr, and Radarr
- DietPi-Software | Fixes
- DietPi-Software | Add space
- DietPi-Software | Add more spaces
- DietPi-Software | Remove unneded line
- DietPi-Software | Fix syntax error
- DietPi-Software | Indent
- DietPi-Software | Increase timeout time
- DietPi-Software | Raise Bazarr timeout
- DietPi-Software | Remove other unneeded line

nonnorm committed:
Add Config Creation function

nonnorm committed:
CodeFactor Fixes

nonnorm committed:
More CodeFactor Fixes

nonnorm committed:
Even more CodeFactor fixes

nonnorm committed:
Fix G_DIETPI_VERSION?

MichaIng commented:
Many thanks!

What do you think about adding this to the `dietpi-software` script instead and then we can replace the per-software code as well with the same PR.
Since I don't see more use then for install scripts, let's not overload DietPi-Globals too much.

nonnorm commented:
Sounds good.  Other than Bazarr and Paper, would this currently be useful for anything else?

nonnorm committed:
Move to DietPi-Software

nonnorm committed:
Use CONFIG_CREATE in Bazarr, Sonarr, and Radarr

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -30,6 +30,29 @@ Available commands:
+  [[ $G_PROGRAM_NAME ]] || local G_PROGRAM_NAME='G_CONFIG_CREATE'
+  local file=$1
+  local service=$2
+  local timeout=15 
```
That timeout should be optionally configurable, since different services might take different startup times, e.g. via optional `$3`?

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -30,6 +30,29 @@ Available commands:
+  if [[ ! -f "$file" ]]
+  then
+    G_DIETPI-NOTIFY 1 "Waiting for {$service^} config file to be created timed out, skipping pre-configuration"
+}
```
A closing `fi` is missing here.

Also: We need an explicit return code when the file has not been created in the end (timed out) so that configuration steps can be done or skipped conditionally.

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -11723,43 +11746,14 @@ _EOF_
-				then
-					G_DIETPI-NOTIFY 1 'Waiting for Bazarr config file to be created timed out, skipping Sonarr and Radarr pre-configuration'
+				CONFIG_CREATE /mnt/dietpi_userdata/bazarr/config/config.ini bazarr
 				else
```
The `else` here is wrong now, also the opening `if` is missing. With an explicit return code (see above), we could do:
```
if CONFIG_CREATE /mnt/dietpi_userdata/bazarr/config/config.ini bazarr
then
```

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -11723,43 +11746,14 @@ _EOF_
-						if [[ ! -f '/mnt/dietpi_userdata/sonarr/config.xml' ]]
-						then
-							G_DIETPI-NOTIFY 1 'Sonarr config file creation timed out, skipping Sonarr pre-configuration'
+							CONFIG_CREATE /mnt/dietpi_userdata/sonarr/config.xml sonarr
```
The conditional structure does not work anymore like this:
If the file already exist **or** has been successfully pre-created, the below pre-configuration shall be done.
So it could be:
```
if [[ -f '/mnt/dietpi_userdata/sonarr/config.xml' ]] || CONFIG_CREATE /mnt/dietpi_userdata/sonarr/config.xml sonarr
then
```

There is another thing:
- The file does not only need to be just created, but the `ApiKey` needs to be present, which is done after a few seconds.
- We need to add an optional content check as well, e.g. via variable ($3 would be optional as timeout): `CONFIG_CREATE_CONTENT='ApiKey'`
- Then in the function the check could be:
  ```
  local content=$CONFIG_CREATE_CONTENT
  unset -v CONFIG_CREATE_CONTENT
  ...
  until [[ -f $file ]] && { [[ ! $content]] || grep -q "$content" "$file"; } || [[ $i -ge $timeout ]] 
  ...
  [[ -f $file ]] && { [[ ! $content]] || grep -q "$content" "$file"; } && return 0

  G_DIETPI-NOTIFY 1 "Waiting for {$service^} config file to be created timed out, skipping pre-configuration"
  return 1
  ```

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -11775,21 +11769,7 @@ _EOF_
 					then
 						if [[ ! -f '/mnt/dietpi_userdata/radarr/config.xml' ]]
 						then
```
Actually even this initial existence check could be added to the function as well, so:
1. Check once if file and in case content is present, before starting the service. If so, `return 0` directly.
2. Start service, wait for file and in case content.
3. Final check if file and content is present.

It's three times the relatively long check block in the function, but overall code is reduced most and it is always both checked, file and content. Currently, if the file exist prior to install, but e.g. it's an old version or the service was started but terminated quickly, the content could be missing, leading to the following configuration steps failing.

nonnorm committed:
Fixes

nonnorm committed:
Add space

nonnorm committed:
Add more spaces

nonnorm committed:
Remove unneded line

nonnorm committed:
Fix syntax error

nonnorm committed:
Indent

nonnorm committed:
Increase timeout time

nonnorm committed:
Raise Bazarr timeout

nonnorm committed:
Remove other unneeded line

nonnorm commented:
VM install and testing was successful, the PR is ready to be merged.

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -30,6 +30,34 @@ Available commands:
+    local content=$CONFIG_CREATE_CONTENT 
+    unset -v CONFIG_CREATE_CONTENT 
+    if [[ -f $file ]] && { [[ ! $content ]] || grep -q "$content" "$file"; } && return 0
+		then
```
Did you review the test installs output? Because this can actually not work. The condition itself contains a return, hence the config file is never pre-created 🙂. Let me help with the syntax.

MichaIng committed:
v6.34

+ DietPi-Software | Create_Config function, rename to match other function names (i.e. Create_User), fix syntax and align indentation

MichaIng commented:
Okay I fixed the syntax. Also I renamed to function to `Create_Config` and variable to `CREATE_CONFIG_CONTENT`, mostly because there is another function `Create_User`, so those are similar now 🙂.

Also the function allows us to reduce the code levels/indentation where it is used, looks much cleaner now 😃.

Will run some final tests, then merge.

MichaIng commented:
#### Fresh install
```
[  OK  ] DietPi-Software | chown -R bazarr:dietpi /mnt/dietpi_userdata/bazarr /opt/bazarr /var/log/bazarr
[  OK  ] DietPi-Software | Starting Bazarr to pre-create config file
[  OK  ] DietPi-Software | systemctl stop bazarr
[  OK  ] DietPi-Software | Starting Sonarr to pre-create config file
[  OK  ] DietPi-Software | systemctl stop sonarr
[  OK  ] DietPi-Software | sed -i /\[sonarr\]/,/^$/s/^ip = .*$/ip = 127.0.0.1/ /mnt/dietpi_userdata/bazarr/config/config.ini
[  OK  ] DietPi-Software | sed -i /\[sonarr\]/,/^$/s/^port = .*$/port = 8989/ /mnt/dietpi_userdata/bazarr/config/config.ini
[  OK  ] DietPi-Software | sed -i /\[sonarr\]/,/^$/s/^base_url = .*$/base_url = \// /mnt/dietpi_userdata/bazarr/config/config.ini
[  OK  ] DietPi-Software | sed -i /\[sonarr\]/,/^$/s/^apikey = .*$/apikey = 7caed586730a42d3af50b4fdd9400e4a/ /mnt/dietpi_userdata/bazarr/config/config.ini
[  OK  ] DietPi-Software | Setting in /mnt/dietpi_userdata/bazarr/config/config.ini adjusted: use_sonarr = True
[  OK  ] DietPi-Software | Starting Radarr to pre-create config file
[  OK  ] DietPi-Software | systemctl stop radarr
[  OK  ] DietPi-Software | sed -i /\[radarr\]/,/^$/s/^ip = .*$/ip = 127.0.0.1/ /mnt/dietpi_userdata/bazarr/config/config.ini
[  OK  ] DietPi-Software | sed -i /\[radarr\]/,/^$/s/^port = .*$/port = 7878/ /mnt/dietpi_userdata/bazarr/config/config.ini
[  OK  ] DietPi-Software | sed -i /\[radarr\]/,/^$/s/^base_url = .*$/base_url = \// /mnt/dietpi_userdata/bazarr/config/config.ini
[  OK  ] DietPi-Software | sed -i /\[radarr\]/,/^$/s/^apikey = .*$/apikey = e1eff6e0b59246b38c33f1d4267b4777/ /mnt/dietpi_userdata/bazarr/config/config.ini
[  OK  ] DietPi-Software | Setting in /mnt/dietpi_userdata/bazarr/config/config.ini adjusted: use_radarr = True
```
👍
#### Reinstall
```
[  OK  ] DietPi-Software | chown -R bazarr:dietpi /mnt/dietpi_userdata/bazarr /opt/bazarr /var/log/bazarr
[  OK  ] DietPi-Software | sed -i /\[sonarr\]/,/^$/s/^ip = .*$/ip = 127.0.0.1/ /mnt/dietpi_userdata/bazarr/config/config.ini
[  OK  ] DietPi-Software | sed -i /\[sonarr\]/,/^$/s/^port = .*$/port = 8989/ /mnt/dietpi_userdata/bazarr/config/config.ini
[  OK  ] DietPi-Software | sed -i /\[sonarr\]/,/^$/s/^base_url = .*$/base_url = \// /mnt/dietpi_userdata/bazarr/config/config.ini
[  OK  ] DietPi-Software | sed -i /\[sonarr\]/,/^$/s/^apikey = .*$/apikey = 7caed586730a42d3af50b4fdd9400e4a/ /mnt/dietpi_userdata/bazarr/config/config.ini
[  OK  ] DietPi-Software | Desired setting in /mnt/dietpi_userdata/bazarr/config/config.ini was already set: use_sonarr = True
[  OK  ] DietPi-Software | sed -i /\[radarr\]/,/^$/s/^ip = .*$/ip = 127.0.0.1/ /mnt/dietpi_userdata/bazarr/config/config.ini
[  OK  ] DietPi-Software | sed -i /\[radarr\]/,/^$/s/^port = .*$/port = 7878/ /mnt/dietpi_userdata/bazarr/config/config.ini
[  OK  ] DietPi-Software | sed -i /\[radarr\]/,/^$/s/^base_url = .*$/base_url = \// /mnt/dietpi_userdata/bazarr/config/config.ini
[  OK  ] DietPi-Software | sed -i /\[radarr\]/,/^$/s/^apikey = .*$/apikey = e1eff6e0b59246b38c33f1d4267b4777/ /mnt/dietpi_userdata/bazarr/config/config.ini
[  OK  ] DietPi-Software | Desired setting in /mnt/dietpi_userdata/bazarr/config/config.ini was already set: use_radarr = True
```
👍

Ready from my end, @ravenclaw900 any issues with my last commit or further tests you want to do?

nonnorm commented:
Looks good to me.  I’ll integrate the function into my Paper PR.  It should be ready later today.

MichaIng commented:
I'll merge this now, so you can merge the current dev branch back to yours.
